File: week07/25/QuickSort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution
{
    public:
    //Function to sort an array using quick sort algorithm.
    void quickSort(int arr[], int low, int high)
    {
        // code here
        if (low < high)
        {
            // Partitioning index, arr[pi] is now at the right place.
            int pi = partition(arr, low, high);

            // Recursively sort elements before partition and after partition
            quickSort(arr, low, pi - 1);  // Sort the left half
            quickSort(arr, pi + 1, high); // Sort the right half
        }
    }
    
    public:
    int partition (int arr[], int low, int high)
    {
       // Your code here
        int pivot = arr[high]; // Choose the last element as pivot
        int i = (low - 1); // Index of smaller element
        
        for (int j = low; j <= high - 1; j++)
        {
            // If current element is smaller than the pivot
            if (arr[j] < pivot)
            {
                i++; // Increment index of smaller element
                std::swap(arr[i], arr[j]);
            }
        }
        std::swap(arr[i + 1], arr[high]); // Move pivot to the right place
        return (i + 1); // Return the partitioning index
    }
};

int main(void) {
    Solution a;
    int hello4[10] = {4,1,3,9,7,2,1};
    
    a.quickSort(hello4, 0, 5);
    for (int i=0; i<7; i++) {
        cout << hello4[i] << " ";
    }
    
    return 1;}

File: week07/29/quick_sort.py
def quickSort(arr,low,high):
    if low<high:
        pi = partition(arr,low,high)  

        quickSort(arr,low,pi-1)
        quickSort(arr,pi+1,high)


def partition(arr,low,high):  # pivot을 기준으로 나눔
    pivot_idx = high
    arr[low], arr[pivot_idx] = arr[pivot_idx], arr[low]  # 첫 번째로 이동
    pivot = arr[low]
    i = low + 1  # 피벗보다 작은 값이 배치될 위치를 추적

    for j in range(low + 1, high + 1):  # pivot 이후부터 끝까지
        if arr[j] < pivot:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1  # 다음 피벗보다 작은 값이 들어갈 위치를 준비
    
    arr[low], arr[i-1] = arr[i-1], arr[low]  # 피벗을 i-1 위치 (pivot보다 작은 마지막 값)에 놓음
    return i-1   # 현재 피벗의 위치


arr = [1, 1, 2, 3, 4, 6, 7, 9, 10]
quickSort(arr,0,len(arr)-1)
print(arr)
